### ISSUE TITLE
DB setting according to postgres jdbc sslmode=prefer

### ISSUE THREAD

**Issue by tingeltangelthomas** [Unleash/unleash#1138]
Hi,
is there any possibility to configure unleash like the postgres jdbc sslmode defaults [1]?
jdbc default is sslmode=prefer: 

> I don't care about encryption but will pay the overhead of encryption if the server supports it

It must use ssl, if available. If not available, connect without encryption.
Unleash must accept self signed certificates with this setting.

Regards
Thomas

[1] https://jdbc.postgresql.org/documentation/head/ssl-client.html

**Comment by ivarconr** [comment]
Hi,

Unleash supports this via the database ssl option, see [the docs](https://docs.getunleash.io/deploy/configuring_unleash#further-customization). 

It is passed directly to the underlying node-postgres module:
https://node-postgres.com/features/ssl

**Comment by tingeltangelthomas** [comment]
For my self signed certificate this helps:
DATABASE_SSL:  { "rejectUnauthorized": false }  

A real "prefer" mode like postgres uses it, is not available as I see in node-postgres repo [1]:
> sslmode=prefer, sslmode=require, sslmode=verify-ca, sslmode=verify-full - sets ssl to true

sslmode=prefer will cause an exception, if db does not support ssl:
```
[2021-12-06T09:58:17.369] [ERROR] server-impl.js - Failed to migrate db OperationalError: The server does not support SSL connections                                                                                                                                                                                          
    at Socket.<anonymous> (/unleash/node_modules/pg/lib/connection.js:71:37)                                                                                                                                                                                                                                                   
    at Object.onceWrapper (events.js:520:26)                                                                                                                                                                                                                                                                                   
    at Socket.emit (events.js:400:28)                                                                                                                                                                                                                                                                                          
    at addChunk (internal/streams/readable.js:293:12)                                                                                                                                                                                                                                                                          
    at readableAddChunk (internal/streams/readable.js:267:9)                                                                                                                                                                                                                                                                   
    at Socket.Readable.push (internal/streams/readable.js:206:10)                                                                                                                                                                                                                                                              
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23) {                                                                                                                                                                                                                                                             
  cause: Error: The server does not support SSL connections
      at Socket.<anonymous> (/unleash/node_modules/pg/lib/connection.js:71:37)
      at Object.onceWrapper (events.js:520:26)
      at Socket.emit (events.js:400:28)
      at addChunk (internal/streams/readable.js:293:12)
      at readableAddChunk (internal/streams/readable.js:267:9)
      at Socket.Readable.push (internal/streams/readable.js:206:10)
      at TCP.onStreamRead (internal/stream_base_commons.js:188:23),
  isOperational: true
}
[ERROR] Error: The server does not support SSL connections
    at Socket.<anonymous> (/unleash/node_modules/pg/lib/connection.js:71:37)
    at Object.onceWrapper (events.js:520:26)
    at Socket.emit (events.js:400:28)
    at addChunk (internal/streams/readable.js:293:12)
    at readableAddChunk (internal/streams/readable.js:267:9)
    at Socket.Readable.push (internal/streams/readable.js:206:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)

```

[1] https://github.com/brianc/node-postgres/blob/4b229275cfe41ca17b7d69bd39f91ada0068a5d0/packages/pg-connection-string/README.md?plain=1#L72

**Comment by ivarconr** [comment]
I think that readme is a bit out of date. Today node-posgres leans in the underlying TCP modules of node.js. Setting `ssl=true` is no longer an option in node.js, but it is essentially the same as setting `ssl={ "rejectUnauthorized": false }`. 

The meaning of postgres "prefer", as far as I understands it, is to use encryption if the server supports it, but I do not care about validating the CA or anything, and that's exactly what `{ "rejectUnauthorized": false }` also do. 

If you are dealing with Self-Signed certificates you have two options:

1. Do not require known validated certificates, this is achieved by setting `DATABASE_SSL: { "rejectUnauthorized": false }`. 
2. Provide the required certificates, either via `DATABASE_SSL` or you can provide a list of CA certificates directly to the Node.js process or the underlying OS. 

What am I missing here?

**Comment by ivarconr** [comment]
After digging more in to this it seems to be some issues with the db-migrate module. 

Steps to reproduce: 
1. Disable ssl on postgres
2. set `DATABASE_SSL={ "rejectUnauthorized": false }`

Stack trace:
```
[2021-12-07T09:57:04.681] [ERROR] server-impl.js - Failed to migrate db OperationalError: The server does not support SSL connections
    at Socket.<anonymous> (/home/ivarconr/code/unleash/node_modules/pg/lib/connection.js:71:37)
    at Object.onceWrapper (events.js:520:26)
    at Socket.emit (events.js:400:28)
    at addChunk (internal/streams/readable.js:293:12)
    at readableAddChunk (internal/streams/readable.js:267:9)
    at Socket.Readable.push (internal/streams/readable.js:206:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
From previous event:
    at dbmigrate.up (/home/ivarconr/code/unleash/node_modules/db-migrate/api.js:202:20)
    at migrateDb (/home/ivarconr/code/unleash/dist/migrator.js:20:16)
    at start (/home/ivarconr/code/unleash/dist/lib/server-impl.js:133:44)
    at /home/ivarconr/code/unleash/dist/server-dev.js:8:39
    at processTicksAndRejections (internal/process/task_queues.js:81:21) {
  cause: Error: The server does not support SSL connections
      at Socket.<anonymous> (/home/ivarconr/code/unleash/node_modules/pg/lib/connection.js:71:37)
      at Object.onceWrapper (events.js:520:26)
      at Socket.emit (events.js:400:28)
      at addChunk (internal/streams/readable.js:293:12)
      at readableAddChunk (internal/streams/readable.js:267:9)
      at Socket.Readable.push (internal/streams/readable.js:206:10)
      at TCP.onStreamRead (internal/stream_base_commons.js:188:23),
  isOperational: true
}
```

Other remarks:
- It does however seem to work if we disable migration (so `knex` and the `pg` module is working just fine). 
- It also handles SSL with self-signed certificates

**Comment by ivarconr** [comment]
Investigating further, it seems like the limiting factor is the underlying node-pg module, directly depending on the `tls.connect` options. As far as I can understand, there is no way this module is going to support sslmode=prefer out of the box. 

From https://node-postgres.com/announcements#2020-02-25 we have: 

![image](https://user-images.githubusercontent.com/158948/145105006-70a59a38-4a23-49b2-a733-163492fd6043.png)

Basically this means we cannot support the "sslmode=prefer" mode out of the box. If you know of the postgres server is going to support SSL or not you might work around it by:

1. Postgres server support SSL with self-signed certificate, set `DATABASE_SSL={ "rejectUnauthorized": false }`
2. Postgres server does not support SSL, set `DATABASE_SSL=false`

**Comment by ivarconr** [comment]
There is a possible workaround in code as well, if you can run Unleash via node.js, but I am not sure if this is something you should run in production. 

```js
const unleash = require('unleash-server');

const config = {
    db: {
        user: 'unleash_user',
        password: 'passord',
        host: 'localhost',
        port: 5432,
        database: 'unleash',
        ssl: { rejectUnauthorized: false },
        disableMigration: false,
    },
}

process.nextTick(async () => {
    try {
        console.log('start Unleash with SSL');
        await unleash.start(config);
    } catch (error) {
        if (error.message.includes('The server does not support SSL connections')) {
            console.log('Start Unleash without SSL');
            await unleash.start({db: {...config.db, ssl: false }})
        }
    }
}, 0);
```

**Comment by ivarconr** [comment]
Also from the [docs](https://www.postgresql.org/docs/current/libpq-ssl.html):

> The default value for sslmode is prefer. As is shown in the table, this makes no sense from a security point of view, and it only promises performance overhead if possible. It is only provided as the default for backward compatibility, and is not recommended in secure deployments.
